### PR TITLE
CIR-955: Added decimal place validation to numerical values

### PR DESCRIPTION
**Pull request by DevAndyroo**

### CODE REVIEW

File: app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala [hunk 1]
```diff
@@ -125,6 +128,12 @@ case class AllocatedRestrictionNegative(i: Int, amt: BigDecimal)(implicit topPat
   val value = Json.obj()
 }
 
+case class AllocatedRestrictionDecimalError(i: Int, amt: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ s"disallowanceAp$i"
+  val errorMessage: String = s"disallowanceAp$i has greater than the allowed 2 decimal places."
+  val value = Json.obj()
```
Comment (burns238) [inline app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala:134]: We want to return the amount in this value

File: app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala [hunk 2]
```diff
@@ -137,6 +146,12 @@ case class AllocatedRestrictionTotalNegative(amt: BigDecimal)(implicit topPath:
   val value = Json.obj()
 }
 
+case class AllocatedRestrictionTotalDecimalError(amt: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ "totalDisallowances"
+  val errorMessage: String = "totalDisallowances has greater than the allowed 2 decimal places."
+  val value = Json.obj()
```
Comment (burns238) [inline app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala:152]: Same here

**Review by burns238 — COMMENTED** [review]
My only comment is to add the values to the error responses - the actual messages will get reviewed in the content review.

File: test/v1/validation/fullReturn/UkCompanyValidatorSpec.scala [hunk 3]
```diff
@@ -63,17 +63,29 @@ class UkCompanyValidatorSpec extends BaseValidationSpec {
         leftSideError(ukCompanyModelReactivationMax.copy(netTaxInterestExpense = -1).validate(groupAccountingPeriod)).errorMessage shouldBe NetTaxInterestExpenseError(-1).errorMessage
       }
 
+      "netTaxInterestExpense is >2 DP" in {
+        leftSideError(ukCompanyModelReactivationMax.copy(netTaxInterestExpense = 2.222, netTaxInterestIncome = 0).validate(groupAccountingPeriod)).errorMessage shouldBe NetTaxInterestExpenseDecimalError(2.222).errorMessage
```
Comment (AdamCoops) [inline test/v1/validation/fullReturn/UkCompanyValidatorSpec.scala:67]: Here

File: app/v1/validation/abbreviatedReturn/AbbreviatedReturnValidator.scala [hunk 4]
```diff
@@ -120,3 +124,9 @@ case class NegativeAngieError(amt: BigDecimal) extends Validation {
   val path = JsPath \ "angie"
   val value = Json.toJson(amt)
 }
+
+case class AngieDecimalError(amt: BigDecimal) extends Validation {
+  val errorMessage: String = "ANGIE has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/abbreviatedReturn/AbbreviatedReturnValidator.scala:129]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/AdjustedGroupInterestValidator.scala [hunk 5]
```diff
@@ -63,8 +68,26 @@ trait AdjustedGroupInterestValidator extends BaseValidation {
...
+
+case class GroupRatioDecimalError(groupEBITDA: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val errorMessage: String = "groupRatio has greater than the allowed 5 decimal places."
+  val path = topPath \ "groupRatio"
+  val value = Json.toJson(groupEBITDA)
+}
+
 case class GroupRatioError(groupRatio: BigDecimal)(implicit topPath: JsPath) extends Validation {
-  val errorMessage: String = "Group Ratio must be between 0 and 100%, and to two decimal places"
+  val errorMessage: String = "Group Ratio must be between 0 and 100%"
```
Comment (versatil56) [inline app/v1/validation/fullReturn/AdjustedGroupInterestValidator.scala:90]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/AllocatedReactivationsValidator.scala [hunk 6]
```diff
@@ -43,3 +45,9 @@ case class AllocatedReactivationsCannotBeNegative(currentPeriodReactivation: Big
   val errorMessage: String = s"currentPeriodReactivation cannot be negative"
   val value = Json.toJson(currentPeriodReactivation)
 }
+
+case class AllocatedReactivationsDecimalError(currentPeriodReactivation: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ "currentPeriodReactivation"
+  val errorMessage: String = s"currentPeriodReactivation is only allowed to be two decimal places"
```
Comment (versatil56) [inline app/v1/validation/fullReturn/AllocatedReactivationsValidator.scala:51]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala [hunk 7]
```diff
@@ -125,6 +128,12 @@ case class AllocatedRestrictionNegative(i: Int, amt: BigDecimal)(implicit topPat
   val value = Json.obj()
 }
 
+case class AllocatedRestrictionDecimalError(i: Int, amt: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ s"disallowanceAp$i"
+  val errorMessage: String = s"disallowanceAp$i has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala:133]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala [hunk 8]
```diff
@@ -137,6 +146,12 @@ case class AllocatedRestrictionTotalNegative(amt: BigDecimal)(implicit topPath:
   val value = Json.obj()
 }
 
+case class AllocatedRestrictionTotalDecimalError(amt: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ "totalDisallowances"
+  val errorMessage: String = "totalDisallowances has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/fullReturn/AllocatedRestrictionsValidator.scala:151]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/FullReturnValidator.scala [hunk 9]
```diff
@@ -238,6 +243,12 @@ case object InterestReactivationCapNotSupplied extends Validation {
   val value = Json.obj()
 }
 
+case class TotalReactivationsDecimalError(amt: BigDecimal) extends Validation {
+  val errorMessage: String = s"totalReactivation has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/fullReturn/FullReturnValidator.scala:247]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/FullReturnValidator.scala [hunk 10]
```diff
@@ -250,6 +261,12 @@ case class TotalReactivationsNotGreaterThanCapacity(calculated: BigDecimal, capa
   val value = Json.obj()
 }
 
+case class TotalRestrictionsDecimalError(amt: BigDecimal) extends Validation {
+  val errorMessage: String = s"totalRestrictions has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/fullReturn/FullReturnValidator.scala:265]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/UkCompanyValidator.scala [hunk 11]
```diff
@@ -101,12 +131,24 @@ case class NetTaxInterestExpenseError(netTaxInterestExpense: BigDecimal)(implici
   val value = Json.toJson(netTaxInterestExpense)
 }
 
+case class NetTaxInterestExpenseDecimalError(netTaxInterestExpense: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ "netTaxInterestExpense"
+  val errorMessage: String = "The supplied Net Tax Interest Expense has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/fullReturn/UkCompanyValidator.scala:136]: This error messages need to feed from messages.en

File: app/v1/validation/fullReturn/UkCompanyValidator.scala [hunk 12]
```diff
@@ -101,12 +131,24 @@ case class NetTaxInterestExpenseError(netTaxInterestExpense: BigDecimal)(implici
...
+
 case class NetTaxInterestIncomeError(netTaxInterestIncome: BigDecimal)(implicit topPath: JsPath) extends Validation {
   val path = topPath \ "netTaxInterestIncome"
   val errorMessage: String = "The supplied Net Tax Interest Income is Negative, which it can not be."
   val value = Json.toJson(netTaxInterestIncome)
 }
 
+case class NetTaxInterestIncomeDecimalError(netTaxInterestIncome: BigDecimal)(implicit topPath: JsPath) extends Validation {
+  val path = topPath \ "netTaxInterestIncome"
+  val errorMessage: String = "The supplied Net Tax Interest Income has greater than the allowed 2 decimal places."
```
Comment (versatil56) [inline app/v1/validation/fullReturn/UkCompanyValidator.scala:148]: This error messages need to feed from messages.en

**Review by versatil56 — CHANGES_REQUESTED** [review]

File: app/v1/validation/abbreviatedReturn/AbbreviatedReturnValidator.scala (hunk 4 above)
Comment (DevAndyroo) [reply app/v1/validation/abbreviatedReturn/AbbreviatedReturnValidator.scala:129]: messages.en doesn't exist on API

**Review by burns238 — APPROVED** [review]

**Pull request merged**